### ISSUE TITLE
Leverage the `--workspace` flag when running cargo command against cargo itself

### ISSUE THREAD

**Issue by weihanglo** [rust-lang/cargo#11987]
### Problem

After #11851 Cargo becomes a Cargo workspace. That means we can dogfood ourselves finally 🎉.

We left some TODOs in that pull request, one of them are making `--workspace` usable for cargo itself. That is, making `cargo build` and other subcommands happy with `--workspace` flag, instead of using `-p` flag everywhere in our CI workflow yaml.

Some workspace members are platform-dependant inside cargo workspace, i.e., `cargo-credential-gnome-secret` and friends. We don't really have a good programming way to get the list other than `cargo metadata | jq .workspace_members[] -r | cut -d' ' -f1`. As a consequence, we expands the list manually in a ugly way: https://github.com/rust-lang/cargo/blob/5861176b5607c67e44ea7d438954fbb5fa20b190/.github/workflows/main.yml#L117-L128

We want to make running command like `cargo test --workspace` as easy as possible to reduce the friction in CI also in contributor experience.

- [ ] ~~Fix clippy warnings and re-enable clippy check in CI for all workspace members.~~
  - Blocked on rust-lang/rfcs#3389 so we can more easily propagate our clippy settings
- [ ] Fix rustdoc warnings and re-enable rustdoc check in CI for all workspace members.
- [ ] Fix `linkchecker.sh` warnings in CI (https://github.com/rust-lang/cargo/pull/11851#discussion_r1147051184)
- [ ] Leverage workspace flag `--workspace` when running `cargo build` or `cargo test`, instead of using flag `-p`.
- [ ] Leverage glob syntax when probing members in `[workspace]` in Cargo.toml (i.e., `crates/*`).

### Proposed Solution

Could look into <https://github.com/rust-lang/cargo/issues/5220> and also <https://github.com/rust-lang/cargo/issues/11526>.

An alternative workaround would be making those subcrates and dependencies platform specific. For example, dependencies could be listed in `target."cfg(not(windows))".dependencies` section. And have a `lib.rs` doing nothing but delegating to `unix/lib.rs` or `windows/lib.rs`, and one of them also does nothing.

### Notes

I think at least the follow commands should be able to run with `--workspace` for cargo itself.

- [ ] `cargo build`/`cargo check`
- [ ] `cargo test`
- [ ] `cargo doc`
- [ ] `cargo clippy` (this is blocked on rust-lang/rfcs#3389, so we can more easily propagate our clippy settings)

**Comment by weihanglo** [comment]
Totally valid. I was waiting for a response from #11526 but no. Closed.

And yep, clarified the short-term and long-term fixes in the PR descriptions. Thanks for the reply!

**Comment by weihanglo** [comment]
> When you say target, this is talking about platform targets, correct?

Yes. Please read relevant issues to track their progress.

This pull request is more like blocked on other pull requests or features. For exammple, I don't think there is an easy way to skip ``cargo-credential-gnome-secret` ` without explicit exclusion, like `cargo test --workspace --exclude "cargo-credential-gnome-secret"`. I don't think there is much we can do at this moment. Maybe only focus on reducing CI yaml? Anyway, feel free to drop your idea.

**Comment by weihanglo** [comment]
Thanks for being interested in this!  However `required-targets` is just one of the  proposed solutions out there. We haven't decided on a proper way to solve this issue.

Note that this is labeled as `A-infrastructure`, meaning that it focuses on making cargo the project infra better with existing functionality in cargo. If we want to resolve this issue with new features, please discuss in their own issues respectively, or open a new issue for a new feature proposal.

Sorry for the confusion 😔.
